semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -485,7 +493,25 @@ let rec m_name a b =
       (* this will handle metavariables in Id *)
       m_ident_and_id_info (a1, a2) (b1, b2)
   | G.Id ((str, tok), _info), G.IdQualified _ when MV.is_metavar_name str ->
-      envf (str, tok) (MV.N b)
+      (* If `is_resolved` is true, then we got the target `IdQualified` from a
+         `ResolvedName`.
+         Such an identifier has a nonsensical range. If the resolved name is `A.foo`, in
+         the following code:
+         ```
+         class A:
+           def foo():
+             pass
+         ```
+         then the pattern metavariable will match the range of the text
+         ```
+         A:
+         def foo
+         ```
+         because it is matching everything between `A` and `foo`.
+
+         So we should not permit this match.
+      *)
+      if is_resolved then fail () else envf (str, tok) (MV.N b)
```

Comment:
Have we considered other solutions?

Reply:
We cannot intercept it by `metavariable-pattern` time, or any time once the binding has been produced, because we will not know whether this originated from a `ResolvedName`. One solution is just to remove the error message entirely, which will lead to `metavariable-pattern` failing, which is good, because then no match will be produced (as it should), but it means _without_ a `metavariable-pattern`, we will produce an extraneous match, which seems to "go away" on a `metavariable-pattern`, regardless of the clause. This does not seem to make sense to me.

Another thing we could do is make an effort to also just fail to produce a match if the expression we are binding does not have a range. This is done in `text_of_binding`, at `Eval_generic` time, but this seems to run later than `envf`, and possibly the match which is being `metavariable-pattern`ed on, which is why we still produce an error. We could add some logic to refuse to `envf` if the expression in question does not have location data. I wanted to restrict it more to this case though, since I don't know what that looks like.

Reply:
It's worth noting that even if we add the latter option (refusing to match nonsensical ranges), this will not work in the grand scheme of my solution for fixing the problems that I see. This is because I intend to make `ResolvedName`s contain location data where possible, and if we produce a match to a `ResolvedName` with true location data, really bad stuff will happen. So I am in favor of recognizing that resolved names behave a little differently, and their location data cannot really be trusted.